File: Anytime_Planning/brentsc-infoplanner/test/integration/test_info_gathering_ARVI.cpp
// Author: Brent Schlotfeldt
// Usage:
//    ../bin/test_info_planner_SE2 ../data/test_one_robot_ARVI.yaml
//
// Profile:
//  Compile with Debug info
//  valgrind --tool=callgrind ../bin/test_info_planner_SE2 ../data/test_one_robot_ARVI.yaml
//  kcachegrind callgrind.out.6000
//
//

#include <igl/env/env_se2.h>
#include <igl/planning/infoplanner.h>
#include <igl/mapping/map_nx.h>
#include <igl/utils/utils_nx.h>
#include <iostream>
#include <yaml-cpp/yaml.h>
#include <Eigen/Dense>
#include <igl/robot.h>
#include <igl/params.h>
#include <igl/estimation/kalman_filter.h>
#include <igl/se3_pose.h>
#include <igl/estimation/multi_target_filter.h>

using namespace Eigen;

int main(int argc, char **argv) {

  nx::Parameters p(argv[1]); // Load parameters from the file given.

  auto robots = p.GetRobots();
  auto planner = p.GetPlanner();
  auto world = p.GetTMM();

  std::vector<std::vector<VectorXd>> fixed_traj(robots.size());
  std::vector<nx::PlannerOutput<nx::SE3Pose>> plannerOutputs(robots.size());
  // Main Loop

  auto t1 = nx::tic();

  for (int t = 0; t < p.Tmax; t += 1) {
    std::cout <<"Timestep: " << t << std::endl;

    // Sense, and Filter.
    for (int i = 0; i < robots.size(); i++) {
      std::vector<nx::Measurement> m = robots[i].sensor->senseMultiple(robots[i].getState(), world);
      nx::GaussianBelief output = nx::MultiTargetFilter<nx::SE3Pose>::MultiTargetKF(m, robots[i]);
      robots[i].tmm->updateBelief(output.mean, output.cov);
    }

    // PlanARVI (Every n_controls steps)
    if (t % p.n_controls == 0) {
      std::cout << "Starting InfoPlanner..." << std::endl;
      // Reset planning variables
      for (int i = 0; i < robots.size(); i++) {
        fixed_traj[i].clear();
      }
      for (int i = 0; i < robots.size(); i++) {
        plannerOutputs[i] = planner->PlanARVI(robots[i], 12);
        // Record fixed trajectories for Coordinate Descent
        int t_ = 0;
        for (auto it = plannerOutputs[i].path.begin();
             it != plannerOutputs[i].path.end(); ++t_, ++it) {
          VectorXd point = robots[i].env->state_to_SE2(*it);
          //std::cout << "robot i: " << point << std::endl;
          fixed_traj[i].push_back(point);
        }
      }
      std::cout << "Computation done in " << nx::toc(t1) << " sec!" << std::endl;
      // TODO Here we should update the collision map, to avoid robots crossing trajectories.
    }

    // Actuate
    for (int i = 0; i < robots.size(); i++) {
      // Apply controls
      robots[i].applyControl(plannerOutputs[i].action_idx, 1);
    }
    // Environment Changes
    world->forwardSimulate(1);
  }

  std::cout << "Final ground truth is: " << world->getTargetState() << std::endl;

  for (int i = 0; i < robots.size(); i++) {
    std::cout << "Final Estimate_" << i << " is: " << robots[i].tmm->getTargetState() << std::endl
              << "Final cov is: "
              << robots[i].tmm->getCovarianceMatrix() << std::endl;
    std::cout << "Final pose_" << i << " is: " << robots[i].getState().position << std::endl;
  }
  return 0;
}

File: Anytime_Planning/brentsc-infoplanner/test/integration/test_info_gathering_ARVI.py
import yaml
import numpy as np
from include.igl.params import Parameters
from include.igl.env.env_se2 import EnvSE2
from include.igl.planning.infoplanner import InfoPlanner
from include.igl.mapping.map_nx import MapNX
from include.igl.utils.utils_nx import tic, toc
from include.igl.robot import Robot
from include.igl.estimation.kalman_filter import KalmanFilter
from include.igl.estimation.multi_target_filter import MultiTargetFilter
from include.igl.se3_pose import SE3Pose

if __name__ == '__main__':
    # Replace 'config_file' with the path to your configuration file
    config_file = 'config_file.yaml'
    params = Parameters(config_file)

    robots = params.get_robots()
    planner = params.get_planner()
    world = params.get_tmm()

    fixed_traj = [[] for _ in range(len(robots))]
    planner_outputs = [None] * len(robots)

    t1 = tic()

    for t in range(params.Tmax):
        print(f"Timestep: {t}")

        # Sense and Filter
        for i, robot in enumerate(robots):
            measurements = robot.sensor.sense_multiple(robot.get_state(), world)
            output = MultiTargetFilter.multi_target_kf(measurements, robot)
            robot.tmm.update_belief(output.mean, output.cov)

        # PlanARVI (Every n_controls steps)
        if t % params.n_controls == 0:
            print("Starting InfoPlanner...")

            # Reset planning variables
            for i in range(len(robots)):
                fixed_traj[i].clear()

            for i, robot in enumerate(robots):
                planner_outputs[i] = planner.plan_arvi(robot, 12)

                # Record fixed trajectories for Coordinate Descent
                for state in planner_outputs[i].path:
                    point = robot.env.state_to_se2(state)
                    fixed_traj[i].append(point)

            print(f"Computation done in {toc(t1)} sec!")

            # TODO: Update the collision map to avoid robots crossing trajectories

        # Actuate
        for i, robot in enumerate(robots):
            robot.apply_control(planner_outputs[i].action_idx, 1)

        # Environment Changes
        world.forward_simulate(1)

    print(f"Final ground truth is: {world.get_target_state()}")

    for i, robot in enumerate(robots):
        print(f"Final Estimate_{i} is: {robot.tmm.get_target_state()}")
        print(f"Final cov is: {robot.tmm.get_covariance_matrix()}")
        print(f"Final pose_{i} is: {robot.get_state().position}")
